table/tables/partition.go

```diff
@@ -204,7 +263,22 @@ func (t *partitionedTable) locatePartition(ctx sessionctx.Context, pi *model.Par
 		// The data does not belong to any of the partition?
 		return 0, errors.Trace(table.ErrTrgInvalidCreationCtx)
 	}
-	return pi.Definitions[idx].ID, nil
+	return idx, nil
+}
+
+// TODO: supports linear hashing
+func (t *partitionedTable) locateHashPartition(ctx sessionctx.Context, pi *model.PartitionInfo, r []types.Datum) (int, error) {
+	ret, isNull, err := t.partitionExpr.Expr.EvalInt(ctx, chunk.MutRowFromDatums(r).ToRow())
+	if err != nil {
+		return 0, err
+	}
+	if isNull {
```

Comment:
Null value in here is reasonable？If not, return error.

Reply:
NULL value is insert into the first partition for the range partition case.
For hash, it's hash partition 0.
This behavior confirms to MySQL document.

Reply:
It is reasonable. 
mysql:
```sql
mysql root@127.0.0.1:test> create table th (
                        ->   a int,
                        ->   b int
                        -> ) partition by hash(a) partitions 5;
Query OK, 0 rows affected
Time: 0.022s
mysql root@127.0.0.1:test> insert into th set b=2;
Query OK, 1 row affected
Time: 0.001s
mysql root@127.0.0.1:test> select * from th;
+--------+---+
| a      | b |
+--------+---+
| <null> | 2 |
+--------+---+
1 row in set
Time: 0.004s
mysql root@127.0.0.1:test>
```